packages: kernel: add kmod-hwmon-emc2305 support

remittor opened the pull request:
package hwmon's `emc2305.ko`. This module supports the Microchip Technology Inc (SMSC) EMC2301/EMC2302/EMC2303/EMC2305 fan speed PWM controllers chips.

This PR is needed to add a new device `Xiaomi R3D`: https://github.com/openwrt/openwrt/pull/14373

motolav commented:
It'd make more sense to move the patches to generic than duplicate them in ipq806x

remittor commented:
> It'd make more sense to move the patches to generic than duplicate them in ipq806x

My understanding is correct that I should do the following?
1) Remove `emc2305` patches from: `./target/linux/bcm27xx/patches-6.1/`
2) Add these patches to: `./target/linux/generic/XXX-6.1`

Subdirectory XXX should be what?  `backport`, `hack` or `pending` ?
**Note:** Kernel 5.15 does not have `emc2305` driver.

Ansuel commented:
@remittor if it has been merged upstream it should be backport it's ok to put patch only in 6.1 if affected target are only 6.1... just make sure to add the related dependency for linux 6.1

remittor commented:
@Ansuel , I didn't really understand what you advised me.

mrkiko commented:
What IPQ6X are you targeting (have a FAN and so on)?

motolav commented:
kernel 6.1 includes the emc2305 driver and you should see if the 6.1 driver can work without any patches

remittor commented:
> What IPQ6X are you targeting (have a FAN and so on)?

I wrote in the first post that I want to add support for router Xiaomi R3D: https://github.com/openwrt-xiaomi/openwrt/commits/master/
The same fan controller is available in router Xiaomi AX9000, but Robimarco was too lazy to control the fan speed.

> and you should see if the 6.1 driver can work without any patches

But these patches have quite significant changes. And I want to make the firmware with the corrected driver.

motolav commented:
> But these patches have quite significant changes. And I want to make the firmware with the corrected driver.

No one has made an effort to upstream the patches so it can't be considered a `backport` or `pending` patch, until it's fixed upstream it's better just to use rudimentary driver the kernel already has

remittor commented:
> it's better just to use rudimentary driver the kernel already has

Will not work. Among these patches there is a critical bug fix: https://github.com/raspberrypi/linux/commit/4cf53e91f36a8e8b7b6ffe33fd5c8ad195ed0f80
In my router `Xiaomi R3D`, the `emc2301` controller is installed.

remittor commented:
> it's better just to use rudimentary driver the kernel already has

It's sad to hear this.
You don’t want to duplicate patches in the `ipq80xx` branches. You don’t want to add to the `generic` branch either....

Ansuel commented:
> It's sad to hear this.

Thing is that it's better to have patch that overlay on top of the upstream driver than having a complete different driver.

It's ok to have downstream patch but if the driver is present upstream in a "rudimentary" state, then it's not acceptable to have a complete different one and we ask to just rework the patch and extend what we have currently.

In these case, if the change would affect multiple target then generic pending patch is needed. If the patch require lots of hack or lots of magic values hack directory is the place to propose the patch.

Hope this is now clear. I notice the patch has been dropped, can someone update me on this?

Ansuel commented:
I'm not following all the soft toxicity here but aside from this... the 3 patch (we can ignore the DT patch) are honestly very simple and they should be very easy to move from bcm to pending-6.1

The fact that ax9000 have the same pwm controller makes it even better for this change.

The confusing part is why this hasn't have been submitted upstream. I need to check where is the discussion for the DT.

rmandrad commented:
i can confirm the kernel modules works without patches ... I have been using on my own build. However the only way to be recognised is to have a dts entry

I am using the following on the ipq8072-ax9000.dts 

&blsp1_i2c6 {
        status = "okay";

        pinctrl-0 = <&i2c_pins>;
        pinctrl-names = "default";

/*        lm63@4C {
               compatible = "ti,lm63";
                reg = <0x4c>;
                #address-cells = <0x01>;
                #size-cells = <0x00>;
        }; */

        fanctrl: emc2301@2f {
                compatible = "smsc,emc2301", "emc2301", "microchip,emc2301";
                reg = <0x2f>;
                #cooling-cells = <2>;
        };

};

robimarko commented:
Other than lack of DT support, this looks OK to me

remittor commented:
@Ansuel 
> The confusing part is why this hasn't have been submitted upstream. I need to check where is the discussion for the DT.

So maybe this PR should be added to the master branch?
If the maintainers decide to transfer driver patches from branch `bcm27xx` to branch `generic`, then this can be done later in another pull request.

Why do I need to wait a colossal amount of time if my pull request can’t break anything?

robimarko commented:
Please make this kmod depend on not using kernel 5.15 as this driver doesn't exist in 5.15 and it will break buildbots for targets using 5.15 otherwise.

remittor committed:
packages: kernel: add kmod-hwmon-emc2305 support

package hwmon's emc2305.ko. This module supports the
Microchip Technology Inc (SMSC) EMC2301/EMC2302/EMC2303/EMC2305
fan speed PWM controllers chips.

Signed-off-by: Oleg S <remittor@gmail.com>

rmandrad commented:
i honestly don't understand the delay... it is an upstream kernel mod ! with so much focus in keeping  openwrt adhered to upstream (and rightly so) what is this about ? @remittor  already added the depend not to include 5.15  so ?

robimarko commented:
@remittor Can you enable editing by maintainers?

I would like to get this merged

remittor commented:
> Can you enable editing by maintainers?

How can I do this?

robimarko commented:
@remittor 
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

remittor commented:
@robimarko 
This checkbox is missing
<details><summary>screenshot</summary>
![image](https://github.com/openwrt/openwrt/assets/9350115/f0f01744-1918-4564-a6d4-6c5a28544951)
</details>

robimarko commented:
Ok, I pushed it manually.

Merged in: https://github.com/openwrt/openwrt/commit/d668c74fe6efb546f0fd65f0334a3f0bd7bf5df4
